File: array_of_products/array_of_products.go
package main

import "fmt"

func arrayOfProducts(array []int) []int {
    products := make([]int, len(array))
    for i := range products {
        products[i] = 1
    }

    product := 1
    for i := len(array) - 1; i >= 0; i-- {
        products[i] = product
        product *= array[i]
    }

    product = 1
    for i := 0; i < len(array); i++ {
        products[i] *= product
        product *= array[i]
    }

    return products
}

func main() {
    result := arrayOfProducts([]int{5, 1, 4, 2})
    fmt.Println(result)
}

File: array_of_products/array_of_products.java
public class ArrayOfProducts {

    public static int[] arrayOfProducts(int[] array) {
        int[] products = new int[array.length];
        for (int i = 0; i < products.length; i++) {
            products[i] = 1;
        }

        int product = 1;
        for (int i = array.length - 1; i >= 0; i--) {
            products[i] = product;
            product *= array[i];
        }

        product = 1;
        for (int i = 0; i < array.length; i++) {
            products[i] *= product;
            product *= array[i];
        }

        return products;
    }

    public static void main(String[] args) {
        int[] result = arrayOfProducts(new int[] {5, 1, 4, 2});
        for (int num : result) {
            System.out.print(num + " ");
        }
    }
}
